File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/init/BaseIntegrator.java
package io.hcxprotocol.init;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import io.hcxprotocol.impl.HCXIncomingRequest;
import io.hcxprotocol.impl.HCXOutgoingRequest;
import io.hcxprotocol.interfaces.IncomingRequest;
import io.hcxprotocol.interfaces.OutgoingRequest;
import io.hcxprotocol.utils.Constants;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;

public class BaseIntegrator {
    private Config config;

    private static Logger logger = LoggerFactory.getLogger(BaseIntegrator.class);

    private void validateConfig() throws Exception {
        logger.debug("Integrator SDK configuration.", getConfig());
        if(config == null)
            throw new Exception("Please initialize the configuration variables, in order to initialize the SDK");
        List<String> props = config.getStringList("configKeys");
        for(String prop: props){
            if(!config.hasPathOrNull(prop) || StringUtils.isEmpty(config.getString(prop)))
                throw new Exception(prop + " is missing or has empty value, please add to the configuration.");
        }
    }

    protected void setConfig(Map<String, Object> map) throws Exception {
        Config fallbackConfig = ConfigFactory.load();
        this.config = ConfigFactory.parseMap(map).withFallback(fallbackConfig);
        validateConfig();
    }

    protected Config getConfig(){
        return this.config;
    }
    
    /**
     * This method is to get the hcx protocol base path.
     */
    public String getHCXProtocolBasePath() {
        return config.getString(Constants.PROTOCOL_BASE_PATH);
    }

    /**
     * This method is to get the participant code.
     */
    public String getParticipantCode() {
        return config.getString(Constants.PARTICIPANT_CODE);
    }

    /**
     * This method is to get the authorization base path.
     */
    public String getAuthBasePath() {
        return config.getString(Constants.AUTH_BASE_PATH);
    }

    /**
     * This method is to get the username.
     */
    public String getUsername() {
        return config.getString(Constants.USERNAME);
    }

    /**
     * This method is to get the password.
     */
    public String getPassword() {
        return config.getString(Constants.PASSWORD);
    }

    /**
     * This method is to get the encryption private key.
     */
    public String getPrivateKey() {
        return config.getString(Constants.ENCRYPTION_PRIVATE_KEY);
    }

    /**
     * This method is to get the HCX implementation guide base path.
     */
    public String getHCXIGBasePath() {
        return config.getString(Constants.HCX_IG_BASE_PATH);
    }

    /**
     * This method is to get the NRCES implementation guide base path.
     */
    public String getNRCESIGBasePath() {
        return config.getString(Constants.NRCES_IG_BASE_PATH);
    }
}

File: dot-net/hcx-integrator-sdk/Io.HcxProtocol/Init/BaseIntegrator.cs
using Io.HcxProtocol.Impl;
using Io.HcxProtocol.Interfaces;
using System;

namespace Io.HcxProtocol.Init
{
    public class BaseIntegrator
    {
        private Config config;

        public void ValidateConfig(Config _config)
        {
            string prop = null;

            if (string.IsNullOrEmpty(_config.ProtocolBasePath))
            {
                prop = "ProtocolBasePath";
            }
            else if (string.IsNullOrEmpty(_config.ParticipantCode))
            {
                prop = "ParticipantCode";
            }
            else if (string.IsNullOrEmpty(_config.ParticipantGenerateToken))
            {
                prop = "ParticipantGenerationToken";
            }
            else if (string.IsNullOrEmpty(_config.UserName))
            {
                prop = "UserName";
            }
            else if (string.IsNullOrEmpty(_config.EncryptionPrivateKey))
            {
                prop = "EncryptionPrivateKey";
            }
            else if (string.IsNullOrEmpty(_config.SigningPrivateKey))
            {
                prop = "SigningPrivateKey";
            }

            if (string.IsNullOrEmpty(_config.HcxIGBasePath))
            {
                //Default Hcx IG Base Path 
                _config.HcxIGBasePath = "https://ig.hcxprotocol.io/v0.7.1/";
            }
            if (string.IsNullOrEmpty(_config.NrcesIGBasePath))
            {
                //Default Nrces IG Base Path 
                _config.NrcesIGBasePath = "https://nrces.in/ndhm/fhir/r4/";
            }
            if (!string.IsNullOrWhiteSpace(_config.LogType) && _config.LogType.ToLower() == "file")
            {
                if (string.IsNullOrEmpty(_config.LogFileName))
                {
                    prop = "LogFileName";
                }
                if (string.IsNullOrEmpty(_config.LogFilePath))
                {
                    prop = "LogiFilePath";
                }
            }

            if (prop != null)
                throw new System.Exception(prop + " is missing or has empty value, please add to the configuration.");
            else
                config = _config;

            ValidateOptionalFields(config);
        }

        private void ValidateOptionalFields(Config _config)
        {
            if (string.IsNullOrEmpty(_config.Password) && string.IsNullOrEmpty(_config.Secret))
            {
                throw new System.Exception("Access token generation failed. Please provide participant password or user secret in the config");
            }
        }

        protected void SetConfig(Config config)
        {
            ValidateConfig(config);
        }

        protected Config GetConfig()
        {
            return config;
        }

        protected IIncomingRequest GetIncomingRequest()
        {
            return GetProcessRequest<HCXIncomingRequest>(config.IncomingRequestClass, typeof(HCXIncomingRequest));
        }

        protected IOutgoingRequest GetOutgoingRequest()
        {
            return GetProcessRequest<HCXOutgoingRequest>(config.OutgoingRequestClass, typeof(HCXOutgoingRequest));
        }

        private T GetProcessRequest<T>(T classInstance, Type defaultClass) where T : class
        {
            if (classInstance != null)
            {
                return classInstance as T;
            }
            return Activator.CreateInstance(defaultClass) as T;
        }

        /**
        * This method is to get the hcx protocol base path.
        */
        public string GetHCXProtocolBasePath()
        {
            return config.ProtocolBasePath;
        }

        /**
        * This method is to get the participant code.
        */
        public string GetParticipantCode()
        {
            return config.ParticipantCode;
        }

        /**
         * This method is to get the authorization base path.
         */
        public string GetAuthBasePath()
        {
            return config.AuthBasePath;
        }

        /**
         * This method is to get the username.
         */
        public string GetUsername()
        {
            return config.UserName;
        }

        /**
        * This method is to get the password.
        */
        public string GetPassword()
        {
            return config.Password;
        }

        /**
        * This method is to get the encryption private key.
        */
        public string GetPrivateKey()
        {
            return config.EncryptionPrivateKey;
        }

        /**
        * This method is to get the HCX implementation guide base path.
        */
        public string GetHCXIGBasePath()
        {
            return config.HcxIGBasePath;
        }

        /**
         * This method is to get the NRCES implementation guide base path.
         */
        public string GetNRCESIGBasePath()
        {
            return config.NrcesIGBasePath;
        }
    }
}
